File: G_Projects/F_Elevator_System/UtilityClasses/ElevatorController.java
package G_Projects.F_Elevator_System.UtilityClasses;

import java.util.ArrayList;
import java.util.List;

import G_Projects.F_Elevator_System.CommandPattern.ConcreteClasses.ElevatorRequest;
import G_Projects.F_Elevator_System.CommonEnums.Direction;
import G_Projects.F_Elevator_System.SchedulingStrategy.SchedulingStrategy;
import G_Projects.F_Elevator_System.SchedulingStrategy.ConcreteStrategies.ScanSchedulingStrategy;

// Manages and coordinates multiple elevators within the building
/**
 * ElevatorController - Central Management System
 * 
 * Key Responsibilities:
 * • Manages multiple elevators within a building system
 * • Initializes and coordinates elevator fleet operations
 * • Processes external elevator requests from floors
 * • Handles internal floor requests from elevator passengers
 * • Implements configurable scheduling strategies for optimal routing
 * • Simulates real-time elevator movement and positioning
 * • Provides centralized control for building-wide elevator operations
 */
public class ElevatorController {
    private List<Elevator> elevators; //List of elevators in the system
    private List<Floor> floors;
    private SchedulingStrategy schedulingStrategy; // Strategy for assigning requests to elevators
    private int currentElevatorId;

    // To define number of elevators and floors in the system
    public ElevatorController(int numberOfElevators, int numberOfFloors) {
        this.elevators = new ArrayList<>();
        this.floors = new ArrayList<>();
        this.schedulingStrategy = new ScanSchedulingStrategy();
        for(int i=0; i<numberOfElevators; i++){
            elevators.add(new Elevator(i));
        }
        for(int i=0; i<numberOfFloors; i++){
            floors.add(new Floor(i));
        }
    }

    // dynamically set scheduling strategy
    public void setSchedulingStrategy(SchedulingStrategy strategy){
        this.schedulingStrategy = strategy;
    }

    // Handle external elevator request
    public void requestElevator(int elevatorId, int floorNumber, Direction direction){
        System.out.println("External request: Requesting elevator " + elevatorId + " to floor " + floorNumber + " going " + direction);

        Elevator selectedElevator = getElevatorById(elevatorId);

        if(selectedElevator != null){
            // Add request to it
            selectedElevator.addRequest(new ElevatorRequest(elevatorId, floorNumber, false, direction));
        }else{
            System.out.println("ElevatorController: No elevator found with ID " + elevatorId);
        }
    }

    // Handle internal elevator request
    public void requestFloor(int elevatorId, int floorNumber){
        Elevator selectedElevator = getElevatorById(elevatorId);
        System.out.println("ElevatorController, Internal request: Elevator " + elevatorId + " requested to go to floor " + floorNumber);

        Direction direction = floorNumber > selectedElevator.getCurrentFloor() ? Direction.UP : Direction.DOWN;

        // add request to elevator
        selectedElevator.addRequest(new ElevatorRequest(elevatorId, floorNumber, false, direction));
    }

    private Elevator getElevatorById(int id){
        for(Elevator elevator : elevators){
           if(elevator.getElevatorId()==id){
                return elevator;
           }
        }
        return null;
    }

    public void step(){
        for(Elevator elevator : elevators){
            if(!elevator.getRequestQueue().isEmpty()){
                // use scheduling strategy to get next stop
                int nextStop = schedulingStrategy.getNextStop(elevator);

                // move elevator to next stop
                if(elevator.getCurrentFloor() != nextStop){
                    elevator.moveToNextStop(nextStop);
                }
            }
        }
    }

    public List<Elevator> getElevators(){
        return elevators;    
    }

    public List<Floor> getFloors(){
        return floors;
    }

    public void setCurrentElevatorId(int id){
        this.currentElevatorId = id;
    }
}

File: Python/A_Project/C_Elevator_System/UtilityClasses/elevator_controller.py
from CommandPattern.ConcreteClasses.elevator_request import ElevatorRequest
from CommonEnums.direction import Direction
from SchedulingStrategy.ConcreteStrategies.scan_scheduling_strategy import ScanSchedulingStrategy
from UtilityClasses.elevator import Elevator
from UtilityClasses.floor import Floor


class ElevatorController:
    def __init__(self, number_of_elevators: int, number_of_floors: int):
        self.elevators = []
        self.floors = []
        self.scheduling_strategy = ScanSchedulingStrategy()
        self.current_elevator_id = 0

        for index in range(number_of_elevators):
            self.elevators.append(Elevator(index))
        for index in range(number_of_floors):
            self.floors.append(Floor(index))

    def set_scheduling_strategy(self, strategy):
        self.scheduling_strategy = strategy

    def request_elevator(self, elevator_id: int, floor_number: int, direction: Direction):
        direction_name = getattr(direction, "name", str(direction))
        print(
            f"External request: Requesting elevator {elevator_id} "
            f"to floor {floor_number} going {direction_name}"
        )
        selected_elevator = self._get_elevator_by_id(elevator_id)
        if selected_elevator is not None:
            selected_elevator.add_request(
                ElevatorRequest(elevator_id, floor_number, False, direction)
            )
        else:
            print(f"ElevatorController: No elevator found with ID {elevator_id}")

    def request_floor(self, elevator_id: int, floor_number: int):
        selected_elevator = self._get_elevator_by_id(elevator_id)
        if selected_elevator is None:
            print(f"ElevatorController: No elevator found with ID {elevator_id}")
            return
        print(
            f"ElevatorController, Internal request: Elevator {elevator_id} "
            f"requested to go to floor {floor_number}"
        )

        direction = (
            Direction.UP
            if floor_number > selected_elevator.get_current_floor()
            else Direction.DOWN
        )
        selected_elevator.add_request(
            ElevatorRequest(elevator_id, floor_number, True, direction)
        )

    def _get_elevator_by_id(self, elevator_id: int):
        for elevator in self.elevators:
            if elevator.get_elevator_id() == elevator_id:
                return elevator
        return None

    def step(self):
        for elevator in self.elevators:
            if elevator.get_request_queue():
                next_stop = self.scheduling_strategy.get_next_stop(elevator)
                if elevator.get_current_floor() != next_stop:
                    elevator.move_to_next_stop(next_stop)
                else:
                    elevator.complete_request()

    def get_elevators(self):
        return self.elevators

    def get_floors(self):
        return self.floors

    def set_current_elevator_id(self, elevator_id: int):
        self.current_elevator_id = elevator_id
